### PR TITLE
Add repl

### PR DESCRIPTION
**Pull request by hayd**
Rebase and work upon #947. cc @cedric05 

Add tests in form of repl_test.py.
Executes js.
No compiling from ts.

Note: atm the scope of eval is window + window.deno. (I haven't been able to get the tests passing  any of the suggested [alternatives](https://github.com/denoland/deno/pull/998#discussion_r225451336)) @kitsonk

### CODE REVIEW

File: tools/repl_test.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    out, err, code = run_repl("var a = 123;", "console.log(a);deno.exit(0)")
+    assert out == '123\n'
+    assert err == ''
+    assert code == 0
+
+
+def deno_reference_error_test():
+    out, err, code = run_repl("not_a_variable", "deno.exit(0)")
+    assert out == ''
+    assert "DenoError ReferenceError: not_a_variable is not defined" in err
```
[the remark this answers is not in the document]
Comment (hayd) [reply tools/repl_test.py]: Yes, oops. I had DenoError there for debugging!

File: js/repl.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    if (!line) { continue }
+    if (line === ".exit") { break }
+    try {
+      let result = eval.call(window, line);
+      // let result = eval.call(replScope, line);
+      // let result = function () { return eval(line); }.call(replScope);
+      // let result = function (line: string) { return eval(line); }.call(replScope, line);
+      if (result) { console.log(result) };
+    } catch (err) {
+      console.error("DenoError", err);
```
[the remark this answers is not in the document]
Comment (hayd) [reply js/repl.ts]: Maybe this should reuse some logic from elsewhere in deno.

I wondered if it was console.trace, will look

File: js/repl.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  msg.Repl.addPrompt(builder, prompt_);
+  const inner = msg.Repl.endRepl(builder);
+  return [builder, msg.Any.Repl, inner];
+}
+
+function res(baseRes: null | msg.Base): string {
+  assert(baseRes != null);
+  assert(msg.Any.ReplRes === baseRes!.innerType());
+  const inner = new msg.ReplRes();
+  assert(baseRes!.inner(inner) != null);
```
[the remark this answers is not in the document]
Comment (hayd) [reply js/repl.ts]: Hmmm this is copy/pasted.

[the remark this answers is not in the document]
Comment (hayd) [reply js/repl.ts]: Hmmm this is copy/pasted.

File: js/repl.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  const line = inner.line();
+  assert(line !== null);
+  return line || '';
+}
+
+export async function repl_loop() {
+  // const replScope = Object.create(window);
+  // Object.defineProperty(replScope, "deno", { value: deno, enumerable: true });
+  // const replScope = Object.defineProperty("deno", Object.create(window), { value: deno, enumerable: true });
+  // const replScope = {"deno": deno, ...window};
```
[the remark this answers is not in the document]
Comment (hayd) [reply js/repl.ts]: I'm still unable to get it.

**Comment by hayd** [comment]
I've removed my attempt at having a single rustyline Editor instance. It seems fast even when the history file starts with 5k lines.

I am not sure where to put history file. I think the scope change could be made later.

**Comment by ry** [comment]
I apologize for the lack of reviews here. I'm glad you guys are working on this. But I want to figure out if we're going to move the event loop into JS (#1003) before landing any REPL code.

File: src/repl.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    })
+    .map_err(|err|
+      match err {
+        Interrupted => exit(1),
+        err => err
+      })
+    .map_err(|err| deno_error(ErrorKind::Other, err.description().to_string()))
+}
+
+fn start_repl() -> Editor<()> {
```
[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: I tried... A lot, see linked commit at the top / comments. I couldn't make the compiler happy.

[the remark this answers is not in the document]
Comment (bartlomieju) [reply src/repl.rs]: Wouldn't it be executed by 2 separate `deno` processes?

I'll take a look and try to hack something myself

[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: Link: https://github.com/hayd/deno/commit/058281039ad4e5e0020d5e6277b1d8c1fe4bc2f6#r30930088

In an ideal world I think you'd want a mutable map to editor instances (so you could create multiple Editors with different Config/histories)... I attempted to do one globally.

[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: One could imagine supporting user defined repls. So you'd create an Editor in fbs and it'd give you a reference to the particular Editor, when you readline you pass the reference to the Editor you want.

This is something node supports. One example might be (though not node-based) is julia, where the command line has regular julia `> `, help mode `help> ` and shell mode `shell> `, each with different histories.

[the remark this answers is not in the document]
Comment (bartlomieju) [reply src/repl.rs]: Thanks for explanation, it makes sense now. Will try to help out

**Comment by bartlomieju** [comment]
@hayd I managed to get REPLs stored on `IsolateState`, you can check it here: https://github.com/bartlomieju/deno/commit/9888d6f33d9e6a0f7e6280989b9528d376fe89f3

**Comment by hayd** [comment]
@bartlomieju 👍 I've merged it to this branch and rebased, but github is still [playing catch up/down](https://status.github.com/messages).

File: js/repl.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kitsonk) [reply js/repl.ts]:
```suggestion
  // TypeScript does not track asserts, baseRes asserted above, therefore not null here
  assert(baseRes!.inner(inner) != null);
```

**Comment by bartlomieju** [comment]
@hayd awesome 👍 I think we can start figuring out how to compile TS to JS. @kitsonk can you give us a hand here and point into right direction?

**Comment by hayd** [comment]
@ry this is mostly ts side now, originally it was calling out isolate's event_loop but now it's not. So it shouldn't be affected by event loop changes.

 Perhaps it should be:
```
🦕>
```

**Comment by ry** [comment]
@hayd I want to get #1081 in before looking into this one.
And no emoji in the prompt. `"> "` is sufficient.

**Comment by hayd** [comment]
rebased. Build fail due to new third_party commit not in denoland repo.

Strangely #1087 doesn't appear to be an issue any more in the repl (or perhaps I've forgotten how to trigger it).

File: src/repl.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: Sigh! This apparently doesn't work on Windows, eurgh! https://ci.appveyor.com/project/deno/deno/builds/20020213 (I think it was this stuff, bleurgh)

**Comment by piscisaureus** [comment]
@hayd - you can work around the issue with this patch: https://github.com/denoland/deno/commit/461557336ca9db5e927cd0878cf4f1325bfbecfb

cc @ry

**Comment by hayd** [comment]
The build/tests are now green on appveyor + travis.

File: src/repl.rs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+pub struct DenoRepl {
+  pub name: String,
+  editor: Editor<()>,
+  history_file: PathBuf,
+}
+
+impl DenoRepl {
+  pub fn new(name: &String, dir: &DenoDir) -> DenoRepl {
```
[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: The intention was to follow node and to support user defined repl. For example, if we add a debugger that may also want to be a separate repl. I think this becomes more powerful/useful when you can pass config to the Editor.

[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: I think this is pretty cheap to support. Though the history file not being where you want it is annoying... I am wondering if there is an easy way to have both.

[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: It should be okay to keep the name and use .deno/{name}_history.txt, realistically there'll only ever be a few.

File: src/repl.rs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+#[cfg(windows)]
+impl<T: rustyline::Helper> DerefMut for Editor<T> {
+  fn deref_mut(&mut self) -> &mut rustyline::Editor<T> {
+    &mut self.inner
+  }
+}
+
+pub struct DenoRepl {
```
[the remark this answers is not in the document]
Comment (hayd) [reply src/repl.rs]: Linereader.readline 🤢 . hmmm

File: src/isolate.rs [hunk 8]
```diff
@@ -105,6 +109,19 @@ impl IsolateState {
...
+    match repls.entry(name.clone()) {
+      Occupied(entry) => entry.into_mut(),
+      Vacant(entry) => entry.insert(DenoRepl::new(&name, &self.dir)),
+    };
+  }
+
+  pub fn exit_repl(&self, name: String) {
+    let mut repls = self.repls.lock().unwrap();
+    repls.remove(&name).unwrap();
+  }
```
Comment (ry) [inline src/isolate.rs]: I think the LineReader should be in the resources table. These functions should be moved/rewritten to be in the style of the resources table.

File: src/msg.fbs [hunk 9]
```diff
@@ -273,6 +277,24 @@ table ReadlinkRes {
   path: string;
 }
 
+table ReplStart {
+  name: string;
+  // TODO add config
+}
```
Comment (ry) [inline src/msg.fbs]: Add `tableReplStartRes { rid: int }`

File: src/msg.fbs [hunk 10]
```diff
@@ -273,6 +277,24 @@ table ReadlinkRes {
   path: string;
 }
 
+table ReplStart {
+  name: string;
```
Comment (ry) [inline src/msg.fbs]: Remove name. You're using it as an index into the `repls` table, but when you use the resources table, there is no index. So the name isn't needed. 

Add  `history_file: string` to specify where to write the history file in the ReplStart message.

**Comment by bartlomieju** [comment]
@hayd do you need help with requested changes?

**Comment by hayd** [comment]
@bartlomieju yes please, if you could add to resources table, change name to history_file etc. would be very helpful, thanks 👍

**Comment by ry** [comment]
@hayd @bartlomieju I'm made the changes to move it into the resources table myself ... because it's a little tricky. (We're going to refactor the resources table soon - to hopefully be less haphazard.)

**Comment by hayd** [comment]
Awesome! Thanks @ry.

The async behavior is from node (and maybe other languages too), but I agree it's kinda strange.... at least, it's strange if you're logging output.

**Comment by hayd** [comment]
@ry I've addressed the other feedback I think (aside from it still being called Repl).

I suspect we'll need to do async. A cool thing you can try is running the repl with debug:
```
$ ./out/debug/deno -D
DEBUG RS - starting background reactor
DEBUG RS - msg_from_js Start sync true
DEBUG JS - cwd /Users/hayd/OSS/deno
DEBUG JS - args []
DEBUG RS - op_repl_start deno_history.txt
DEBUG RS - Loading REPL history: "/Users/hayd/.deno/deno_history.txt"
DEBUG RS - msg_from_js ReplStart sync true
DEBUG RS - op_repl_readline 3 >
> DEBUG RS - key: Char('a')
DEBUG RS - Emacs command: SelfInsert(1, 'a')
DEBUG RS - Changeset::insert(0, 'a')
> aDEBUG RS - key: Char('b')
DEBUG RS - Emacs command: SelfInsert(1, 'b')
DEBUG RS - Changeset::insert(1, 'b')
> abDEBUG RS - key: Char('c')
DEBUG RS - Emacs command: SelfInsert(1, 'c')
DEBUG RS - Changeset::insert(2, 'c')
> abc
```
Every key stroke logged, yet the repl remains functional... which is nice.

**Comment by ry** [comment]
> I suspect we'll need to do async (at some point).

 Yah, I think so too...

**Pull request merged**